src/github.com/matrix-org/dendrite/encryptoapi/storage/encrypt_keys_table.go

```diff
@@ -0,0 +1,197 @@
+`
+const selectAllkeysSQL = `
+SELECT user_id, device_id, key_id, key_type, key_info, algorithm, signature FROM encrypt_keys 
+WHERE user_id = $1 AND key_type = $2
+`
+
+type keyStatements struct {
+	insertKeyStmt       *sql.Stmt
+	selectKeyStmt       *sql.Stmt
+	selectInKeysStmt    *sql.Stmt
+	selectAllKeyStmt    *sql.Stmt
+	selectSingleKeyStmt *sql.Stmt
+	deleteSingleKeyStmt *sql.Stmt
+}
+
+func (s *keyStatements) prepare(db *sql.DB) (err error) {
+	_, err = db.Exec(keysSchema)
+	if err != nil {
+		return
+	}
+	if s.insertKeyStmt, err = db.Prepare(insertkeySQL); err != nil {
+		return
+	}
+	if s.selectKeyStmt, err = db.Prepare(selectkeySQL); err != nil {
+		return
+	}
+	if s.selectInKeysStmt, err = db.Prepare(selectInkeysSQL); err != nil {
+		return
+	}
+	if s.selectAllKeyStmt, err = db.Prepare(selectAllkeysSQL); err != nil {
+		return
+	}
+	if s.deleteSingleKeyStmt, err = db.Prepare(selectSinglekeySQL); err != nil {
+		return
+	}
+	if s.selectSingleKeyStmt, err = db.Prepare(deleteSinglekeySQL); err != nil {
+		return
+	}
+	return
+}
+
+// insert keys
+func (ks *keyStatements) insertKey(
+	ctx context.Context, txn *sql.Tx,
+	deviceID, userID, keyID, keyTyp, keyInfo, algorithm, signature string,
+) error {
+	stmt := common.TxStmt(txn, ks.insertKeyStmt)
+	_, err := stmt.ExecContext(ctx, deviceID, userID, keyID, keyTyp, keyInfo, algorithm, signature)
+	return err
+}
+
+// select by user and device
+func (ks *keyStatements) selectKey(
+	ctx context.Context,
+	txn *sql.Tx,
+	deviceID, userID string,
+) (holders []types.KeyHolder, err error) {
+	stmt := common.TxStmt(txn, ks.selectKeyStmt)
+	rows, err := stmt.QueryContext(ctx, userID, deviceID)
+	if err != nil {
+		return nil, err
+	}
+	for rows.Next() {
+		single := &types.KeyHolder{}
+		if err := rows.Scan(
+			&single.User_id,
+			&single.Device_id,
+			&single.Key_id,
+			&single.Key_type,
+			&single.Key,
+			&single.Key_algorithm,
+			&single.Signature,
+		); err != nil {
+			return nil, err
+		}
+		holders = append(holders, *single)
+	}
+	return holders, err
+}
+
+// select single one for claim usage
+func (ks *keyStatements) selectSingleKey(
+	ctx context.Context,
+	userID, deviceID, algorithm string,
+) (holder types.KeyHolder, err error) {
+	stmt := ks.selectSingleKeyStmt
+	row := stmt.QueryRowContext(ctx, userID, deviceID, algorithm)
+	if err != nil {
+		return holder, err
+	}
+	if err := row.Scan(
+		&holder.User_id,
+		&holder.Device_id,
+		&holder.Key_id,
+		&holder.Key_type,
+		&holder.Key,
+		&holder.Key_algorithm,
+		&holder.Signature,
+	); err != nil {
+		deleteStmt := ks.deleteSingleKeyStmt
+		deleteStmt.ExecContext(ctx, userID, deviceID, algorithm, holder.Key_id)
+		return holder, err
+	}
+	return holder, err
+}
+
+// select details by given an array of devices
+func (ks *keyStatements) selectInKeys(
+	ctx context.Context,
+	userID string,
+	arr []string,
+) (holders []types.KeyHolder, err error) {
+	rows := &sql.Rows{}
```

Comment:
I believe rows needs to be closed to avoid leaks.

Reply:
yeah , I forgot it